Convert register device into suspendable. Allow to check for FCM registration success.

wzieba opened the pull request:
Closes #2398 

Corresponding Woo PR: https://github.com/woocommerce/woocommerce-android/pull/6535

***Description***
This PR adds a coroutine-based API for registering devices on WPCOM. It also adds the feature of checking if the device is registered.

❗️Please wait with merging this PR until https://github.com/woocommerce/woocommerce-android/pull/6535 is approved!

wzieba committed:
Convert register device into suspendable

wzieba committed:
Add GetDeviceRegistrationStatus

wzieba committed:
Use coroutine engine for register device endpoint

wzieba committed:
Merge branch 'trunk' into issue/2398_register_device_suspendable

wzieba committed:
Return unregistered state if registered device id is null

wzieba commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/GetDeviceRegistrationStatus.kt:
```diff
@@ -0,0 +1,21 @@
+    private val prefsWrapper: PreferenceUtils.PreferenceUtilsWrapper
+) {
+    operator fun invoke(): Status {
+        val deviceId = prefsWrapper.getFluxCPreferences().getString(NotificationStore.WPCOM_PUSH_DEVICE_SERVER_ID, null)
```
When a device gets registered on WPCOM, we receive a "server id" and save it in SharedPreferences.

I think that checking for the presence of this "server id" is the most robust way to check if device registration was indeed successful.

wzieba replied:
When a device gets registered on WPCOM, we receive a "server id" and save it in SharedPreferences.

I think that checking for the presence of this "server id" is the most robust way to check if device registration was indeed successful.

wzieba committed:
Merge branch 'trunk' into issue/2398_register_device_suspendable

atorresveiga commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/notifications/NotificationRestClient.kt:
```diff
@@ -55,7 +55,49 @@ class NotificationRestClient @Inject constructor(
+                        "version_code" to PackageUtils.getVersionCode(appContext).toString(),
+                        "os_version" to Build.VERSION.RELEASE,
+                        "device_uuid" to uuid
+                ).toMap(),
```
Not sure if I'm missing something, but if `mapOf` is returning a map, why do we need to call `toMap()`

wzieba replied:
You're right - it's not needed, thanks! Fixed in 85664cffd1333e67da3646b0e9ffd9ca33d125fc

atorresveiga commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/NotificationStore.kt:
```diff
@@ -273,6 +273,36 @@ class NotificationStore @Inject constructor(
     fun getNotificationByLocalId(noteId: Int) =
             notificationSqlUtils.getNotificationByIdSet(NoteIdSet(noteId, 0, 0))
 
+    suspend fun registerDevice(token: String, appKey: NotificationAppKey): RegisterDeviceResponsePayload {
```
I noticed that we have `WooPayload` and `WooResult` classes. I'm wondering if this refactor is a good opportunity to update the function response? (use WooPayload and WooResult). Is it in our plans to update all results to use these classes? I'm asking to learn because I was in doubt in the past about how and when to use these classes 😁

wzieba replied:
That'd be a good idea to migrate but the device registration code is used by both WordPress and WooCommerce.

This particular method is not yet used, but I think that in future it will be, so I'd leave the `RegisterDeviceResponsePayload` return type, WDYT?

atorresveiga commented on example/src/test/java/org/wordpress/android/fluxc/store/GetDeviceRegistrationStatusTest.kt:
```diff
@@ -0,0 +1,58 @@
+import org.wordpress.android.fluxc.utils.PreferenceUtils
+
+@RunWith(MockitoJUnitRunner::class)
+class GetDeviceRegistrationStatusTest {
```
Excellent work adding tests! 🙌

atorresveiga commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/NotificationStore.kt:
```diff
@@ -273,6 +273,36 @@ class NotificationStore @Inject constructor(
+        }
+    }
+
+    @Deprecated("EventBus is deprecated.", ReplaceWith("registerDevice(token, appKey)"))
```
Nice!

atorresveiga commented on fluxc/src/main/java/org/wordpress/android/fluxc/action/NotificationAction.java:
```diff
@@ -19,6 +19,7 @@
 public enum NotificationAction implements IAction {
     // Remote actions
     @Action(payloadType = RegisterDevicePayload.class)
+    @Deprecated
```
Nice!

atorresveiga reviewed (approved):
Great work Wojtek! I had some questions out of curiosity but the changes looks good to me and the tests are all green so I will pre-approve this 👍

wzieba committed:
Remove redundant toMap()

wzieba commented:
Thanks @atorresveiga ! I've addressed the comments. I'll wait with the merge of this PR until Woo PR will be ready to merge as it might require FluxC changes.

wzieba committed:
Add use case for invalidating device registration

wzieba committed:
Merge branch 'trunk' into issue/2398_register_device_suspendable
